src/osd/osd_types.cc

```diff
@@ -5878,3 +5878,33 @@ ostream& operator<<(ostream& out, const store_statfs_t &s)
       << ")";
   return out;
 }
+
+void OSDOp::clear_data(vector<OSDOp>& ops)
+{
+  for (unsigned i = 0; i < ops.size(); i++) {
+    OSDOp& op = ops[i];
+    op.outdata.clear();
+    if (ceph_osd_op_type_attr(op.op.op) &&
+        op.op.xattr.name_len &&
+	op.indata.length() >= op.op.xattr.name_len) {
+      bufferptr bp(op.op.xattr.name_len);
+      bufferlist bl;
+      bl.append(bp);
+      bl.copy_in(0, op.op.xattr.name_len, op.indata);
+      op.indata.claim(bl);
```

Comment:
> Most of the data is freed, only minimal data kept in memory.

ahh, right! you are copying only the name of xattr, or the cls and method name.

> By using fixed size buffer pointer, and the kept info is just several bytes allocated via tcmalloc/jemalloc, which should be efficient. 
> With the action of copy those several bytes, not rebuilding whole bufferlist.

okay, assuming the names are very short.